### PR TITLE
refactor(node/rolldown): single source of truth for plugin hooks

### PR DESCRIPTION
**Pull request by hyf0**
### Description

This prevent bugs of forgetting define hooks in multiple places.

### CODE REVIEW

**Comment by hyf0** [comment]
* **#1943** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1943?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1942** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1942?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1941** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/1941?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @hyf0 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/rolldown/src/plugin/index.ts [hunk 1]
```diff
@@ -187,27 +204,31 @@ export type AsyncPluginHooks = Exclude<
   SyncPluginHooks
 >
 
-export type FirstPluginHooks =
+export type FirstPluginHooks = DefinedHookNames[
   | 'load'
-  | 'renderDynamicImport'
+  // | 'renderDynamicImport'
```
Comment (hyf0) [inline packages/rolldown/src/plugin/index.ts:209]: In this PR, we catched `renderDynamicImport` is actually not used.

File: packages/rolldown/src/constants/plugin.ts [hunk 2]
```diff
@@ -37,8 +26,59 @@ export const ENUMERATED_PLUGIN_HOOK_NAMES = [
...
+ *   options: 'options',
+ *   buildStart: 'buildStart',
+ *   ...
+ * }
+ * ```
  */
-export const PLUGIN_HOOK_NAMES_SET = new Set(
-  ENUMERATED_PLUGIN_HOOK_NAMES as readonly string[],
-)
+export const DEFINED_HOOK_NAMES: DefinedHookNames = {
```
Comment (IWANABETHATGUY) [inline packages/rolldown/src/constants/plugin.ts:63]: :+1:

**Comment by hyf0** [comment]
### Merge activity

* **Aug 11, 11:51 AM EDT**: [Graphite](https://app.graphite.dev/github/pr/rolldown/rolldown/1942) rebased this pull request after merging its parent, because this pull request is set to merge when ready.

**Pull request merged**